ray-on-gke/tpu/kuberay-tpu-webhook/main.go

```diff
@@ -357,39 +417,79 @@ func getReplicaIndex(clusterName string, groupName string, namespace string) int
 }
 
 // returns next lowest TPU_WORKER_ID in pod slice and updates mappings
-func getNextWorkerID(podSlice slice, namespace string, replicaIndex int) int {
-	tpuWorkerID := 0
-	if sliceToWorkers[podSlice] == nil {
-		newWorker := worker{tpuWorkerID, replicaIndex, true}
-		sliceToWorkers[podSlice] = []worker{newWorker}
-	} else {
-		nextLowestID := math.MaxInt32
-		replacePod := false
-		// iterate through existing workers and check if any have been deleted
-		for _, worker := range sliceToWorkers[podSlice] {
-			if worker.isCreated == false && worker.workerIndex < nextLowestID {
-				replacePod = true
-				nextLowestID = worker.workerIndex
+func getNextWorkerID(sliceToWorkerIDs map[slice][]int, podSlice slice, namespace string, replicaIndex int) int {
+	tpuWorkerID := 0 // defaults to 0 (first Pod in slice)
+	if sliceToWorkerIDs[podSlice] != nil {
+		sort.Ints(sliceToWorkerIDs[podSlice])
+		// iterate through existing workers and get the next lowest, unused ID
+		for _, workerID := range sliceToWorkerIDs[podSlice] {
+			if workerID == tpuWorkerID {
+				tpuWorkerID++
 			}
 		}
-		// reassign next lowest TPU_WORKER_ID if pod has been deleted
-		if replacePod == true {
-			for index, worker := range sliceToWorkers[podSlice] {
-				// set worker.isCreated to true now that pod is being re-created
-				if worker.workerIndex == nextLowestID {
-					sliceToWorkers[podSlice][index].isCreated = true
+	}
+	klog.V(1).InfoS("getNextWorkerID", "RayCluster", namespace+"/"+podSlice.clusterName, "Worker Group", podSlice.groupName, "TPU_WORKER_ID", tpuWorkerID)
+	return tpuWorkerID
+}
+
+// builds mapping representing the current RayCluster state of TPU pods using PodInformer
+func getSliceToWorkerIDs(podsInGroup []*corev1.Pod, clusterName string, groupName string, namespace string, numOfHosts int32) (map[slice][]int, error) {
```

Comment:
Small suggestions to remove indentation in code:
```
// builds mapping representing the current RayCluster state of TPU pods using PodInformer
func getSliceToWorkerIDs(podsInGroup []*corev1.Pod, clusterName string, groupName string, namespace string, numOfHosts int32) (map[slice][]int, error) {
	sliceToWorkerIDs := make(map[slice][]int)

	if podsInGroup == nil {
		// return an empty mapping if no Pods with 'ray.io/group' label found
		return sliceToWorkerIDs, nil
	}
	klog.V(1).InfoS("getSliceToWorkerIDs", "RayCluster", namespace+"/"+clusterName, "# Pods in Group", len(podsInGroup))
	for _, existingPod := range podsInGroup {
		if existingPod.Status.Phase != "Pending" && existingPod.Status.Phase != "Running" {
			continue
		}
		existingClusterName := existingPod.Labels["ray.io/cluster"]
		existingGroupName := existingPod.Labels["ray.io/group"]
		existingNamespace := existingPod.Namespace
		// we only care about workers in the same RayCluster and worker group when assigning IDs
		if clusterName != existingClusterName || groupName != existingGroupName || namespace != existingNamespace {
			continue
		}

		if !containerRequestingTPUs(existingPod.Spec.Containers...) {
			// Pod does not request TPUs, 'ray.io/group' is not a TPU worker group
			return sliceToWorkerIDs, nil
		}
		replicaIndexLabel := existingPod.Labels["replicaIndex"]
		if replicaIndexLabel == "" {
			// Pod has not been intercepted by the KubeRay TPU webhook yet
			continue
		}
		replicaIndexLabelValues := strings.Split(replicaIndexLabel, "-")
		existingReplicaIndex, _ := strconv.Atoi(replicaIndexLabelValues[len(replicaIndexLabelValues)-1])
		existingWorkerID := -1
		for _, container := range existingPod.Spec.Containers {
			if !containerRequestingTPUs(container) {
				continue
			}

			tpuWorkerIDEnvVar := getEnvironmentVariable("TPU_WORKER_ID", container)
			tempVar, err := strconv.Atoi(tpuWorkerIDEnvVar)
			if err != nil {
				klog.ErrorS(err, "getSliceToWorkerIDs", "RayCluster", namespace+"/"+clusterName, "TPU_WORKER_ID", tpuWorkerIDEnvVar)
				continue
			}
			existingWorkerID = tempVar
			break
		}
		if existingPod.Status.Phase == "Running" && existingWorkerID == -1 {
			return nil, errors.New("existing TPU worker missing TPU_WORKER_ID")
		}
		if existingWorkerID != -1 {
			// Pod has been intercepted by the webhook
			podSlice := slice{existingClusterName, existingGroupName, namespace, existingReplicaIndex, numOfHosts}
			if sliceToWorkerIDs[podSlice] == nil {
				sliceToWorkerIDs[podSlice] = []int{existingWorkerID}
			} else {
				sliceToWorkerIDs[podSlice] = append(sliceToWorkerIDs[podSlice], existingWorkerID)
			}
			klog.V(1).InfoS("getSliceToWorkerIDs", "RayCluster", namespace+"/"+clusterName, "ReplicaIndex", existingReplicaIndex, "TPU_WORKER_ID", existingWorkerID)

		}
	}
	return sliceToWorkerIDs, nil
}
```

Reply:
Added this change in [221d595](https://github.com/GoogleCloudPlatform/ai-on-gke/pull/723/commits/221d5950ee8b8387aa43bf63fd115a3ad45a0ad6)